### PR TITLE
Fix abscal tip tilt correction with global phase slope solver and iterative TT_phs_cal

### PR DESCRIPTION
**Pull request by jsdillon**
This PR addresses the issue with how abscal's post-omnical tip-tilt phase correction works, as I explained in #257. 

Along the way, I've also addressed #245, #253, and #254, which are mostly related to standardizing inputs and clarifying documentation in abscal.

### CODE REVIEW

File: hera_cal/abscal_funcs.py [hunk 1]
```diff
@@ -585,6 +585,90 @@ def delay_slope_lincal(model, data, antpos, wgts=None, df=9.765625e4, medfilt=Tr
 
     return fit
 
+def global_phase_slope_logcal(model, data, antpos, wgts=None, verbose=True, tol=1.0):
+    """
+    Solve for a frequency-independent spatial phase slope using the equation 
+
+    median_over_freq(V_ij,xy^data / V_ij,xy^model) = dot(Phi_x, r_i) - dot(Phi_y, r_j)
```
Comment (nkern) [inline hera_cal/abscal_funcs.py]: is there supposed to be an `angle` somewhere on the LHS of this eqn?

File: hera_cal/abscal_funcs.py [hunk 2]
```diff
@@ -506,7 +506,7 @@ def delay_slope_lincal(model, data, antpos, wgts=None, df=9.765625e4, medfilt=Tr
     -------
     fit : dictionary containing delay slope (T_x) for each pol [seconds / meter].
     """
-    echo("...configuring linsolve data for delay_lincal", verbose=verbose)
+    echo("...configuring linsolve data for delay_slope_lincal", verbose=verbose)
```
Comment (nkern) [inline hera_cal/abscal_funcs.py]: thanks for the fix.. i've been meaning to fix this specifically for months... haha

File: hera_cal/abscal_funcs.py (hunk 1 above)
Comment (jsdillon) [reply hera_cal/abscal_funcs.py]: Good call. Fixed.

File: hera_cal/abscal_funcs.py [hunk 3]
```diff
@@ -585,6 +585,90 @@ def delay_slope_lincal(model, data, antpos, wgts=None, df=9.765625e4, medfilt=Tr
...
+    
+    # build linear system
+    ls_data, ls_wgts = {}, {}
+    for rk in red_keys:
+        # build equation string 
+        eqn_str = '{}*Phi_ew_{} + {}*Phi_ns_{} - {}*Phi_ew_{} - {}*Phi_ns_{}'
+        eqn_str = eqn_str.format(antpos[rk[0]][0], rk[2][0], antpos[rk[0]][1], rk[2][0],
+                                 antpos[rk[1]][0], rk[2][1], antpos[rk[1]][1], rk[2][1])
+
+        # calculated frequency median of unflagged angle(data/model)
```
Comment (nkern) [inline hera_cal/abscal_funcs.py]: is it better to average the first data across freq, and then take `angle`?

File: hera_cal/abscal.py [hunk 4]
```diff
@@ -44,17 +44,22 @@ class AbsCal(object):
         where T_dly is a delay slope in [ns / meter]
         and B_ij is the baseline vector between ant i and j.
 
-    5) Average amplitude linear calibration solves the equation:
+    5) frequency-independent phase slope calibration
+        median_over_freq(V_ij^data / V_ij^model) = dot(Phi, B_ji)
```
Comment (nkern) [inline hera_cal/abscal.py]: needs an `angle`

File: hera_cal/abscal_funcs.py (hunk 2 above)
Comment (jsdillon) [reply hera_cal/abscal_funcs.py]: Haha, no problem.

Code reviews are one thing, but I definitely understand abscal a lot better now that I've had to *modify* it.

File: hera_cal/abscal.py [hunk 5]
```diff
@@ -1075,66 +1202,74 @@ def abscal_run(data_files, model_files, calfits_infiles=None, verbose=True, over
...
+            io.write_cal(output_calfits_path, gain_dict, data_freqs, data_times, flags=flag_dict, quality=None, 
+                 total_qual=None, return_uvc=False, overwrite=overwrite, history=history)
 
     # form return tuple
     return_obj = ()
 
     # return gains if desired
     if return_gains:
-        return_obj += (gains,)
+        return_obj += (gain_dict,)
```
Comment (nkern) [inline hera_cal/abscal.py]: why was this changed?

File: hera_cal/tests/test_abscal.py [hunk 6]
```diff
@@ -531,13 +555,13 @@ def test_omni_abscal_arg_parser(self):
         a = hc.abscal.omni_abscal_arg_parser()
 
     def test_abscal_run(self):
```
Comment (nkern) [inline hera_cal/tests/test_abscal.py]: did you add an `abscal_run` call with `phs_slope_cal = True`

File: hera_cal/abscal.py (hunk 5 above)
Comment (jsdillon) [reply hera_cal/abscal.py]: `gains` is no longer defined. It used to hold a list of `gain_dict`s, one per data file. But now there's just one data file.

File: hera_cal/tests/test_abscal.py (hunk 6 above)
Comment (jsdillon) [reply hera_cal/tests/test_abscal.py]: Done.

File: hera_cal/abscal.py (hunk 5 above)
Comment (nkern) [reply hera_cal/abscal.py]: ah, oh yeah that's right.. okay cool

File: hera_cal/abscal_funcs.py (hunk 3 above)
Comment (nkern) [reply hera_cal/abscal_funcs.py]: so what about phase wrapping issues... i.e. when the ratio of data / model isn't that close and oscillates between and around 0 and -2pi?

Comment (jsdillon) [reply hera_cal/abscal_funcs.py]: It's a fair question. Both methods suffer from phase wraps potentially... the hope is that the median fixes it either way as long as it doesn't phase wrap too much.

This is just my experience with one data file, which admittedly, is picked to be very similar to the calibrator. I think it's worth revisiting this if the fix doesn't stop sources from moving around in our images.

Also, the point of this is to just get things closer... the iterative TT will get us the rest of the way if there's lots of phase wrapping.

**Comment by jsdillon** [comment]
Great. Thanks @nkern
